### ISSUE TITLE
Index on docket_number is too big and needs to be made partial

### ISSUE THREAD

**Issue by mlissner** [freelawproject/courtlistener#3543]
## The Scene

I'm setting up a new replication user on postgres 16.1, and I'm getting a sawtooth pattern as shown in #3509. 

## Findings

At first, I thought this was because I didn't have enough logical slots while creating the new subscriber, but I restarted the replication from scratch and that didn't fix the issue, so I started digging into the logs and found this at the peak of a saw tooth:

> 2024-01-01 00:25:31 UTC::@:[29891]:ERROR: index row size 2824 exceeds btree version 4 maximum 2704 for index "search_docket_docket_number_7642c6c6dbd04704_uniq"
> 2024-01-01 00:25:31 UTC::@:[29891]:DETAIL: Index row references tuple (1922780,7) in relation "search_docket".
> 2024-01-01 00:25:31 UTC::@:[29891]:HINT: Values larger than 1/3 of a buffer page cannot be indexed.
> Consider a function index of an MD5 hash of the value, or use full text indexing.
> 2024-01-01 00:25:31 UTC::@:[29891]:CONTEXT: COPY search_docket, line 66527254
> 2024-01-01 00:25:31 UTC::@:[395]:LOG: background worker "logical replication worker" (PID 29891) exited with exit code 1

One theory about this error is that it is due to btree indexes have a maximum size, which got smaller between versions. This is explained in a [Stackoverflow Q&A](https://stackoverflow.com/questions/70123567/index-row-size-2712-exceeds-btree-version-4-maximum-2704-for-index-while-doing), which points to [Postgres documentation](https://www.postgresql.org/docs/12/release-12.html#id-1.11.6.14.4), which says:

> In new btree indexes, the maximum index entry length is reduced by eight bytes, to improve handling of duplicate entries (Peter Geoghegan)

I'm not entirely convinced by this because we're replicating from v14 to v16 and the change happened when v12 was released, but the error is pretty clear. 

In any case, the solution proposed in the log seems fine: Use an MD5 of the big value in the index.

The index from the log, `search_docket_docket_number_7642c6c6dbd04704_uniq`, is defined in the DB as 

```postgres
courtlistener=> \d search_docket_docket_number_7642c6c6dbd04704_uniq
Index "public.search_docket_docket_number_7642c6c6dbd04704_uniq"
    Column     |          Type          | Key? |  Definition   
---------------+------------------------+------+---------------
 docket_number | text                   | yes  | docket_number
 pacer_case_id | character varying(100) | yes  | pacer_case_id
 court_id      | character varying(15)  | yes  | court_id
unique, btree, for table "public.search_docket"
```

In Python, it shows up as:

```python
class Docket(AbstractDateTimeModel):
    class Meta:
        unique_together = ("docket_number", "pacer_case_id", "court")
```

In https://github.com/freelawproject/courtlistener/issues/1406#issuecomment-846292290 we have a query that tells us about bad indexes. I'm happy to say this index doesn't show up there. 

However, if you look at its stats, it's a very big index:

```
courtlistener=>  SELECT indexrelname "Index Name",
 pg_size_pretty(pg_relation_size(indexrelid)) "Index Size"
 FROM pg_stat_all_indexes i JOIN pg_class c ON i.relid=c.oid 
 WHERE i.relname='search_docket' and indexrelname = 'search_docket_docket_number_7642c6c6dbd04704_uniq';
                    Index Name                     | Index Size 
---------------------------------------------------+------------
 search_docket_docket_number_7642c6c6dbd04704_uniq | 4167 MB
```

-----

## Solution

I think the log lines give a good solution: We should implement this unique key as an MD5 hash of the docket number field.[^1]

Django has [an MD5 function these days](https://docs.djangoproject.com/en/5.0/ref/models/database-functions/#md5), so this should be a matter of a fairly simple DB migration to make the unique key a combination of the md5 of the docket number, the pacer ID and the court.

## Open questions:

1. Can this be migrated in place or do we need to create a new index then delete the old. I can't really imagine what would happen if we migrated in place, because if we:
    - **Migrate the publisher first:** We'll send MD5-hashed values to the subscriber before it knows what to do with them.
    - **Migrate the subscriber first:** We'll send unhashed values to the subscriber which will trip it up.

    See more details on adding columns [on our wiki](https://github.com/freelawproject/courtlistener/wiki/Database-migrations#adding-a-column). 

3. Regardless of Q1, once the new index is in place, do we have to restart the sync of the DB that's currently sawtoothing or will it somehow manage (I'm doubtful, but it's not a big deal to restart it).

## Next steps

 - [ ] Make two PRs: One to add the new index, and one to remove the old one. Ensure they're [created `CONCURRENTLY`](https://github.com/freelawproject/courtlistener/wiki/Database-migrations#adding-and-removing-indexes-to-a-column-dont-use-concurrently-by-default).
 - [ ] Migrate all DBs.

[^1]: My gut said to use SHA1 or SHA2 keys here, since MD5 is no good for security, but MD5 is faster, and fine for this purpose. It also doesn't require a special postgres security extension.

**Comment by mlissner** [comment]
@ERosendo, we should get this done as soon as possible. Could you put this on top of your list, please on Tuesday, unless there's something else urgent that I've forgotten?

**Comment by ERosendo** [comment]
> Could you put this on top of your list

@mlissner Sure, I'll review this issue and get it done to you as soon as possible.

**Comment by ERosendo** [comment]
> Can this be migrated in place or do we need to create a new index then delete the old

Adding the new index in two steps(adding the new one first, and then dropping the old index) seems like the best approach.

>  Do we have to restart the sync of the DB that's currently sawtoothing or will it somehow manage?

I don't believe that the replica will be able to resolve the issue as the initial sync was not completed properly. I think the replica will try to copy data from the old publisher snapshot, which includes the problematic index. Restarting seems to be the best solution to fix this problem.

**Comment by mlissner** [comment]
I realized something about the way replication works and I think I disagree. When you do logical replication, it migrates the data as tuples, but it doesn't migrate the indexes — they're built on the subscriber server as the tuples arrive. I think if we simply delete the index on the subscriber, the replication will be fixed. 

We'll still need to create the new, better index, but I bet it will work in the meantime.

I'll try this later today.

**Comment by mlissner** [comment]
OK, after the above, replication finally worked, thank goodness. Let's convert the docket number index as well, as per above, and then we'll be good to go, I think.

**Comment by mlissner** [comment]
This is fixed via #3558, and I'll deploy via #3615. Closing, thank you.
